Account for conflicted files in git_status()

jennybc opened the pull request:
Fixes #39

jennybc committed:
Account for conflicted files in git_status()

jennybc committed:
Add a test for status with a conflicted file

jennybc committed:
Add NEWS bullet

jennybc committed:
Populate PR number

jennybc commented on tests/testthat/test-commit.R:
```diff
@@ -72,3 +72,35 @@ test_that("creating a commit in another directory without author works", {
 })
+
+test_that("status reports a conflicted file", {
+  # temporary measure until gert can do a non fast forward merge
+  skip_if_not_installed("git2r")
```
I know this is a non-starter but I have to be able to *create* a merge conflict and at the moment I can't with gert alone, which is #41.

jeroen replied:
Totally fine as a for now. I don't mind at all adding `git2r` to Suggests, it might be useful for some other cross-validation tests between gert/git2r as well.

jennybc replied:
OK added git2r to Suggests, at least for a little while.

jennybc commented on tests/testthat/test-commit.R:
```diff
@@ -72,3 +72,35 @@ test_that("creating a commit in another directory without author works", {
+  git_commit("Uppercase first 2 words", repo = repo)
+
+  # TODO: gert needs the ability to merge with conflicts
+  git2r::merge(x = repo, y = "my-branch")
```
Once #41 is handled, this becomes some gert call.

jennybc commented on tests/testthat/test-commit.R:
```diff
@@ -72,3 +72,35 @@ test_that("creating a commit in another directory without author works", {
+  git2r::merge(x = repo, y = "my-branch")
+
+  status <- git_status(repo)
+  expect_equal(status$file, "foo.txt")
+  expect_equal(status$status, "conflicted")
+  expect_false(status$staged)
```
I did this column-wise to avoid any data.frame vs. tibble stuff.

jennybc committed:
Suggest git2r

Needed at least in the short-term, in order to create a merge conflict, which is used in a test of git_status().

jennybc committed:
Put a testing signature in setup.R; make other test self-contained

jennybc committed:
Work around lack of configured user

jennybc commented on tests/testthat/setup.R:
```diff
@@ -0,0 +1,4 @@
+gert_testing_signature <- git_signature(
+  name <- 'Testing Jerry',
+  email <- 'test@jerry.com'
+)
```
It would probably be easier to just configure a git user. Either in the CI config files or here (if no existing user is configured). Or write a `local_git_user()` type of helper. Thoughts?

jeroen replied:
CRAN has in the past yelled at me for gert modifying the `~/.gitconfig`. So we could only do it per repo, not globally.

jennybc commented on tests/testthat/test-commit.R:
```diff
@@ -72,3 +71,41 @@ test_that("creating a commit in another directory without author works", {
+  git2r_sig <- structure(
+    list(name = "Testing Jerry", email = "test@jerry.com", when = when),
+    class = "git_signature"
+  )
```
git2r doesn't offer a constructor for its `git_signature` object. It would probably be nicer to configure local user here?

jennybc committed:
Different approach to configuring a user in tests
